### PR TITLE
[7.2.x] JBPM-5531: Problems after deleting a form that is used as a subform in other form.

### PR DESCRIPTION
**Pull request by pefernan**
This PR fixes the issue of removing forms that are used as nested forms. The main changes are:
- New component to display asset usages based on the old ShowUsagesPopup. This component allows any editor to check the usages of the asset or asset part its being edited  (https://github.com/pefernan/kie-wb-common/blob/JBPM-5531-remove-forms/kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/java/org/kie/workbench/common/services/refactoring/client/usages/ShowAssetUsagesDisplayer.java)
- Adapted DataModeller & FormModeller to use it and prevent deletions when assets are being used..
- New checks on Form Editor  to warn the user that the  nested forms have been removed (previously it was breaking)

Since the change affects different areas I'd like to ask different people to review the changes:
- @wmedvede could you please review the changes on Data Modeller?
- @Rikkola could you please review how ShowAssetUsagesDisplayer is querying the index?
- @manstis could you please review the ShowAssetUsagesDisplayer component?
- @jsoltes could you please review the forms part?

This PR is part of an ensemble, please merge with:
https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/490
https://github.com/kiegroup/kie-wb-common/pull/1009
https://github.com/kiegroup/drools-wb/pull/561
https://github.com/kiegroup/optaplanner-wb/pull/192
https://github.com/kiegroup/kie-wb-distributions/pull/563
https://github.com/kiegroup/jbpm-wb/pull/817
https://github.com/kiegroup/jbpm-form-modeler/pull/128

### CODE REVIEW

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/pom.xml [hunk 1]
```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (manstis) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/pom.xml]: License?

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/ErraiApp.properties [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#
+# Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/ErraiApp.properties]: 2017

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/META-INF/beans.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (manstis) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/META-INF/beans.xml]: License

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/screens/ProjectDiagramExplorerScreen.java [hunk 4]
```diff
@@ -147,7 +147,7 @@ public String getMyContextRef() {
     }
 
     public void show(final ClientSession session) {
-        // Do not show sessions not already initialized with some diagram instance.
+        // Do not showAssetPartUsages sessions not already initialized with some diagram instance.
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/screens/ProjectDiagramExplorerScreen.java]: Refactoring gone wrong :-(

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/screens/ProjectDiagramPropertiesScreen.java [hunk 5]
```diff
@@ -123,7 +123,7 @@ private void handleSession(final ClientSession session) {
                         .show(() -> view.hideLoading());
                 done = true;
             } catch (ClassCastException e) {
-                // No writteable session. Do not show properties until read mode available.
+                // No writteable session. Do not showAssetPartUsages properties until read mode available.
```
Comment (manstis) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/screens/ProjectDiagramPropertiesScreen.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java [hunk 6]
```diff
@@ -1068,7 +1015,7 @@ protected void updateSource(String source) {
 
     private void updateEditorView(DataObject dataObject) {
         //here we need to check if data object name, or package, changed, etc.
-        //if this is the likely we can show an alert to the user, etc.
+        //if this is the likely we can showAssetPartUsages an alert to the user, etc.
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/jpadomain/properties/IdGeneratorEditionDialog.java [hunk 7]
```diff
@@ -153,7 +153,7 @@ public void hide() {
 
     @Override
     public String getStringValue() {
-        //return the value to show in the property editor simple text field.
+        //return the value to showAssetPartUsages in the property editor simple text field.
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/jpadomain/properties/IdGeneratorEditionDialog.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/jpadomain/properties/RelationshipEditionDialog.java [hunk 8]
```diff
@@ -312,7 +312,7 @@ public void hide() {
 
     @Override
     public String getStringValue() {
-        //return the value to show in the property editor simple text field.
+        //return the value to showAssetPartUsages in the property editor simple text field.
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/jpadomain/properties/RelationshipEditionDialog.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/jpadomain/properties/SequenceGeneratorEditionDialog.java [hunk 9]
```diff
@@ -182,7 +182,7 @@ public void hide() {
 
     @Override
     public String getStringValue() {
-        //return the value to show in the property editor simple text field.
+        //return the value to showAssetPartUsages in the property editor simple text field.
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/jpadomain/properties/SequenceGeneratorEditionDialog.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-datasource-mgmt/kie-wb-common-datasource-mgmt-client/src/test/java/org/kie/workbench/common/screens/datasource/management/client/editor/datasource/DataSourceDefEditorTest.java [hunk 10]
```diff
@@ -230,7 +230,7 @@ public void testOnBrowseDatabase() {
         //open the editor with a valid content
         prepareLoadFileSuccessful();
 
-        //emulates the selection of the show content action from the UI
+        //emulates the selection of the showAssetPartUsages content action from the UI
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-datasource-mgmt/kie-wb-common-datasource-mgmt-client/src/test/java/org/kie/workbench/common/screens/datasource/management/client/editor/datasource/DataSourceDefEditorTest.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-library/kie-wb-common-library-client/src/test/java/org/kie/workbench/common/screens/library/client/screens/ProjectMetricsScreenTest.java [hunk 11]
```diff
@@ -243,7 +243,7 @@ public void testAlwaysShow7Days() throws Exception {
         DataSet dataSet = presenter.getAllCommitsDisplayer().getDataSetHandler().getLastDataSet();
         assertEquals(dataSet.getRowCount(), 2);
 
-        // Bar chart must always show 7 bars, one per day of week
+        // Bar chart must always showAssetPartUsages 7 bars, one per day of week
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-library/kie-wb-common-library-client/src/test/java/org/kie/workbench/common/screens/library/client/screens/ProjectMetricsScreenTest.java]: Refactoring gone wrong :-(

File: kie-wb-common-screens/kie-wb-common-social-home-page/kie-wb-common-social-home-page-client/src/main/java/org/kie/workbench/common/screens/social/hp/client/homepage/DefaultSocialLinkCommandGenerator.java [hunk 12]
```diff
@@ -93,7 +93,7 @@ void generateNoRightsPopup() {
                 Constants.INSTANCE.Error_NoAccessRights(),
                 null,
                 null,
-                () -> {/* do nothing, just to show the cancel button*/} );
+                () -> {/* do nothing, just to showAssetPartUsages the cancel button*/} );
```
Comment (manstis) [inline kie-wb-common-screens/kie-wb-common-social-home-page/kie-wb-common-social-home-page-client/src/main/java/org/kie/workbench/common/screens/social/hp/client/homepage/DefaultSocialLinkCommandGenerator.java]: Refactoring gone wrong :-(

File: kie-wb-common-forms/kie-wb-common-forms-editor/kie-wb-common-forms-editor-backend/src/main/java/org/kie/workbench/common/forms/editor/backend/indexing/FormDefinitionIndexer.java [hunk 13]
```diff
@@ -88,7 +88,7 @@ protected DefaultIndexBuilder fillIndexBuilder(Path path) throws Exception {
                 FormModelVisitorProvider provider = visitorProviders.get(model.getClass());
                 if (provider != null) {
                     FormModelVisitor modelVisitor = provider.getVisitor();
-                    modelVisitor.index(model);
+                    modelVisitor.index(formDefinition ,model);
```
Comment (manstis) [inline kie-wb-common-forms/kie-wb-common-forms-editor/kie-wb-common-forms-editor-backend/src/main/java/org/kie/workbench/common/forms/editor/backend/indexing/FormDefinitionIndexer.java]: Formatting

File: kie-wb-common-forms/kie-wb-common-forms-editor/kie-wb-common-forms-editor-backend/src/main/java/org/kie/workbench/common/forms/editor/backend/indexing/FormDefinitionVisitor.java [hunk 14]
```diff
@@ -41,5 +43,13 @@ public void visit(FieldDefinition fieldDefinition) {
                          PartType.FORM_FIELD);
         addResourceReference(fieldDefinition.getStandaloneClassName(),
                              ResourceType.JAVA);
+        if(fieldDefinition instanceof HasNestedForm) {
```
Comment (manstis) [inline kie-wb-common-forms/kie-wb-common-forms-editor/kie-wb-common-forms-editor-backend/src/main/java/org/kie/workbench/common/forms/editor/backend/indexing/FormDefinitionVisitor.java]: Formatting

**Comment by manstis** [comment]
@wmedvede Want to review your Data Object Editor changes?

@pefernan Can you please add whoever from QE as a reviewer (I forget who does "Forms")

**Comment by pefernan** [comment]
@manstis already added @jsoltes.

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/META-INF/beans.xml (hunk 3 above)
Comment (pefernan) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/META-INF/beans.xml]: @manstis good point, I never add license headers  to non-java files. But in this case... I always thought that license headers on beans.xml could break the app deployment, obviously I might be wrong.

File: kie-wb-common-forms/kie-wb-common-forms-editor/kie-wb-common-forms-editor-client/src/test/java/org/kie/workbench/common/forms/editor/client/editor/changes/ChangesNotificationDisplayerTest.java [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+import org.kie.workbench.common.forms.editor.service.shared.model.FormModelSynchronizationUtil;
+import org.kie.workbench.common.forms.model.FieldDefinition;
+import org.kie.workbench.common.forms.model.FormDefinition;
+import org.kie.workbench.common.forms.model.ModelProperty;
+import org.kie.workbench.common.forms.model.TypeKind;
+import org.kie.workbench.common.forms.model.impl.ModelPropertyImpl;
+import org.kie.workbench.common.forms.model.impl.TypeInfoImpl;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.omg.CORBA.Object;
```
Comment (Rikkola) [inline kie-wb-common-forms/kie-wb-common-forms-editor/kie-wb-common-forms-editor-client/src/test/java/org/kie/workbench/common/forms/editor/client/editor/changes/ChangesNotificationDisplayerTest.java]: Crime is a disease CORBA is the answer.

**Comment by Rikkola** [comment]
If forms can depend on forms. Is it possible to make A depend on B and B depend on A and then neither can be removed?

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/META-INF/beans.xml (hunk 3 above)
Comment (manstis) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-client/src/main/resources/META-INF/beans.xml]: You might be right! My eyes glaze over when I look at licenses and code formatting...

**Comment by pefernan** [comment]
@Rikkola good point, nowadays there's no control over recursive forms nesting nesting (A -> B -> A) I must add this control (somehow) on the editor to avoid this kind of issues when configuring nested forms.

BTW, the user will be free to remove the form A or B, the modal is just a warning, telling that the form is being used... he will always have make the final decission.

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-backend/src/main/java/org/kie/workbench/common/forms/jbpm/server/service/indexing/TaskFormModelVisitor.java [hunk 16]
```diff
@@ -37,7 +41,7 @@
     }
 
     @Override
-    public void index(TaskFormModel formModel) {
+    public void index(FormDefinition formDefinition, TaskFormModel formModel) {
```
Comment (Rikkola) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-backend/src/main/java/org/kie/workbench/common/forms/jbpm/server/service/indexing/TaskFormModelVisitor.java]: Is this FormDefinition used?

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/AssetsUsageServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+        indexResource("Bike.java");
+        indexResource("Car.java");
+    }
+
+    protected void indexResource(String resource) throws IOException, InterruptedException {
+        final org.uberfire.java.nio.file.Path path = basePath.resolve(resource);
+        final String content = loadText(resource);
+        ioService().write(path,
+                          content);
+        Thread.sleep(5000);
```
Comment (Rikkola) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/AssetsUsageServiceImplTest.java]: At one point I calculated that our test sleeps take 5 minutes, I wonder how much they are now. Well not much we do about it I guess.

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+import org.kie.workbench.common.services.refactoring.backend.server.TestIndexer;
+import org.kie.workbench.common.services.refactoring.backend.server.indexing.AbstractFileIndexer;
+import org.kie.workbench.common.services.refactoring.backend.server.indexing.DefaultIndexBuilder;
+import org.kie.workbench.common.services.refactoring.service.ResourceType;
+import org.kie.workbench.common.services.shared.project.KieProjectService;
+import org.uberfire.backend.server.util.Paths;
+import org.uberfire.io.IOService;
+import org.uberfire.java.nio.file.Path;
+
+public class TestJavaIndexer extends AbstractFileIndexer implements TestIndexer<TestJavaResourceTypeDefinition> {
```
Comment (Rikkola) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java]: I think we already have a Java file indexer here:
https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/indexing/JavaFileIndexer.java

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java [hunk 19]
```diff
@@ -319,37 +314,14 @@ private void onSafeDelete() {
...
-                }
-            }).findClassUsages(versionRecordManager.getPathToLatest(),
-                               className);
+            showAssetUsagesDisplayer.showAssetUsages(Constants.INSTANCE.modelEditor_confirm_deletion_of_used_class(className),
+                                                     versionRecordManager.getCurrentPath(),
+                                                     className,
+                                                     ResourceType.JAVA,
+                                                     () -> onDelete(versionRecordManager.getPathToLatest()),
+                                                     () -> {
+                                                     });
```
Comment (Rikkola) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java]: Looks to me like the last parameter never does anything, but not sure if there is some location that uses it and it is just not in the diff.

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java (hunk 18 above)
Comment (pefernan) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java]: @Rikkola yes, but the I prefer not to add dependencies to other modules like data-modeler at the refactoring level. BTW the one I added is doing some stuff specific for this test like that are not covered by the one that already exist.

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java (hunk 19 above)
Comment (pefernan) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java]: @Rikkola you're right but I'm pretty sure that if I don't add the cancel callback somebody will ask for that ;)

File: kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-backend/src/main/java/org/kie/workbench/common/forms/jbpm/server/service/indexing/TaskFormModelVisitor.java (hunk 16 above)
Comment (pefernan) [inline kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-backend/src/main/java/org/kie/workbench/common/forms/jbpm/server/service/indexing/TaskFormModelVisitor.java]: @Rikkola not on this implementation of FormModelVisitor for example this one is using it (https://github.com/pefernan/kie-wb-common/blob/c454dd67defe6104531668d7254249ab22865de5/kie-wb-common-forms/kie-wb-common-forms-integrations/kie-wb-common-forms-jbpm-integration/kie-wb-common-forms-jbpm-integration-backend/src/main/java/org/kie/workbench/common/forms/jbpm/server/service/indexing/BusinessProcessFormModelVisitor.java#L62)

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/AssetsUsageServiceImplTest.java (hunk 17 above)
Comment (pefernan) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/AssetsUsageServiceImplTest.java]: I think that @hasys did something about this in other tests, I'll take a look. BTW yes, I think that we cannot do much more for the indexing tests.

**Comment by pefernan** [comment]
@manstis better now?

**Comment by pefernan** [comment]
@manstis please don't merge yet... it seems that I forgot another showcase

**Comment by manstis** [comment]
@pefernan I wasn't going to merge yet.. there are reviews outstanding and the full build has failed...

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/AssetsUsageServiceImplTest.java (hunk 17 above)
Comment (Rikkola) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/AssetsUsageServiceImplTest.java]: You can take a quick look if you want to, but probably not worth investing too much time on it. This is just me wondering how much we end up sleeping.

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java (hunk 18 above)
Comment (Rikkola) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java]: @pefernan The issue I see is that if we index in both then project asset file list for example will list the item twice, just checking this does not happen.

Comment (manstis) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java]: @Rikkola This is a ```test``` class.

**Comment by pefernan** [comment]
Jenkins please retest this

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java [hunk 20]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2012 Red Hat, Inc. and/or its affiliates.
- * <p/>
+ * <p>
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java]: @pefernan seems like this html paragraph was added in the past by mistake.
would you mind fix this licence header?

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java [hunk 21]
```diff
@@ -319,37 +314,14 @@ private void onSafeDelete() {
...
-                                    }
-                                });
-                    } else {
-                        //no usages, just proceed with the deletion.
-                        onDelete(versionRecordManager.getPathToLatest());
-                    }
-                }
-            }).findClassUsages(versionRecordManager.getPathToLatest(),
-                               className);
+            showAssetUsagesDisplayer.showAssetUsages(Constants.INSTANCE.modelEditor_confirm_deletion_of_used_class(className),
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java]: @pefernan A test that checks the showAssetUsagesDisplayer will be appreciated please.

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java [hunk 22]
```diff
@@ -434,38 +406,13 @@ private void onSafeRename() {
...
-                                }
-                        );
-                    } else {
-                        //no usages, just proceed with the deletion.
-                        rename();
-                    }
-                }
-            }).findClassUsages(versionRecordManager.getPathToLatest(),
-                               className);
+            showAssetUsagesDisplayer.showAssetUsages(Constants.INSTANCE.modelEditor_confirm_renaming_of_used_class(className),
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenPresenter.java]: @pefernan A test that checks the showAssetUsagesDisplayer is invocked will be appreciated please.

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenViewImpl.java [hunk 23]
```diff
@@ -1,12 +1,12 @@
 /**
  * Copyright 2012 Red Hat, Inc. and/or its affiliates.
- *
+ * <p>
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/DataModelerScreenViewImpl.java]: @pefernan seems like this html paragraph was added in the past by mistake.
would you mind fix this licence header?

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/maindomain/MainDataObjectEditor.java [hunk 24]
```diff
@@ -122,276 +128,259 @@ public void onNameChange() {
...
-                            );
-
-                        } else {
-                            //no usages, just proceed with the class name change.
-                            doClassNameChange( packageName, oldValue, newValue );
-                        }
-                    }
-                } ).findClassUsages( currentPath, originalClassName );
+            if (originalClassName != null) {
+                showAssetUsagesDisplayer.showAssetUsages(Constants.INSTANCE.modelEditor_confirm_renaming_of_used_class(originalClassName),
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/maindomain/MainDataObjectEditor.java]: @pefernan A test that checks the showAssetUsagesDisplayer is invocked will be appreciated please.

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/maindomain/MainDataObjectFieldEditor.java [hunk 25]
```diff
@@ -115,255 +121,258 @@ public Widget asWidget() {
...
-                        );
-
-                    } else {
-                        //no usages, just proceed with the deletion.
-                        doFieldNameChange( oldValue, newValue );
-                    }
-                }
-            } ).findFieldUsages( currentPath, originalClassName, oldValue );
+        if (originalClassName != null) {
+            showAssetUsagesDisplayer.showAssetPartUsages(Constants.INSTANCE.modelEditor_confirm_renaming_of_used_field(oldValue),
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/maindomain/MainDataObjectFieldEditor.java]: @pefernan A test that checks the showAssetUsagesDisplayer is invocked will be appreciated please.

**Comment by wmedvede** [comment]
@pefernan one more simple change. Can you take a look at the KieWorkbenchDatamodellerClient.gwt.xml and KieWorkbenchDatamodellerWithDataSourcesClient.gwt.xml, and add the missing   
<inherits name="org.kie.workbench.common.services.refactoring.KieWorkbenchCommonRefactoringClient"/>
please?
Thanks.

**Comment by wmedvede** [comment]
@pefernan one last silly comment. Probably the "OK" button of the "Usages notification" can be the red danger button, instead of the blue one. Since in general we are warning the user about potential damages. ¿?

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java (hunk 18 above)
Comment (Rikkola) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/test/java/org/kie/workbench/common/services/refactoring/backend/server/query/assetUsages/TestJavaIndexer.java]: And so it is. Well, then no worries :)

**Comment by pefernan** [comment]
@wmedvede fixed your comments. BTW added the Ok button (now Proceed) with the Danger style ;)

**Comment by jsoltes** [comment]
jenkins please execute full downstream build

**Comment by wmedvede** [comment]
jenkins retest this

File: kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/rendering/renderers/relations/multipleSubform/MultipleSubFormFieldRenderer.java [hunk 26]
```diff
@@ -58,9 +58,13 @@ public void initInputWidget() {
         }
         if (field.getCreationForm() == null || field.getCreationForm().isEmpty()) {
             configErrors.add(FormRenderingConstants.MultipleSubformNoCreationForm);
+        } else if (!renderingContext.getAvailableForms().containsKey(field.getCreationForm())) {
```
Comment (pefernan) [inline kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/rendering/renderers/relations/multipleSubform/MultipleSubFormFieldRenderer.java]: @jsoltes yes this one https://issues.jboss.org/browse/JBPM-6019

Comment (jsoltes) [inline kie-wb-common-forms/kie-wb-common-dynamic-forms/kie-wb-common-dynamic-forms-client/src/main/java/org/kie/workbench/common/forms/dynamic/client/rendering/renderers/relations/multipleSubform/MultipleSubFormFieldRenderer.java]: @pefernan thanks!

**Comment by pefernan** [comment]
@jsoltes better now?

**Comment by pefernan** [comment]
jenkins retest this

**Comment by pefernan** [comment]
jenkins please execute full downstream build

**Comment by pefernan** [comment]
OMG I see this one green too... I cannot believe it

**Comment by pefernan** [comment]
@manstis could it be merged?

**Comment by manstis** [comment]
@pefernan I see https://github.com/kiegroup/kie-wb-distributions/pull/563 is still red (and that's just the "local" build, not full downstream). I think we need to get all "local" builds green and one "full downstream" green before we can safely merge.

**Pull request merged**